models/issue_milestone.go

```diff
@@ -311,71 +311,61 @@ func ChangeMilestoneStatus(m *Milestone, isClosed bool) (err error) {
 	return sess.Commit()
 }
 
-func changeMilestoneIssueStats(e *xorm.Session, issue *Issue) error {
-	if issue.MilestoneID == 0 {
-		return nil
-	}
-
-	m, err := getMilestoneByRepoID(e, issue.RepoID, issue.MilestoneID)
-	if err != nil {
-		return err
-	}
-
-	if issue.IsClosed {
-		m.NumOpenIssues--
-		m.NumClosedIssues++
-	} else {
-		m.NumOpenIssues++
-		m.NumClosedIssues--
-	}
+func updateMilestoneTotalNum(e Engine, milestoneID int64) (err error) {
+	_, err = e.Exec("UPDATE `milestone` SET num_issues=(SELECT count(*) FROM issue WHERE milestone_id=?), completeness=if(num_issues>0,100*num_closed_issues/num_issues, 0) WHERE id=?",
```

Comment:
I'm afraid `completeness` will lag behind one issue. It may be engine dependent, but i think that `num_issues` on the right is evaluated before `num_issues` on the left is updated. `completeness` is not currently checked in any tests, AFAIK.

Reply:
To solve the division by 0 without using `if()`, you could use:
```
UPDATE milestone SET num_issues=(SELECT count(*) FROM issue WHERE milestone_id=?),
completeness = 100*num_closed_issues/(num_issues+1-sign(num_issues)) WHERE id=?;
```

Reply:
However it does lag one value behind:
```
; First, a milestone with wrong value in num_issues:
giteatest=> select id, name, num_issues, num_closed_issues, completeness
    from milestone where id = 1;
 id |      name       | num_issues | num_closed_issues | completeness
----+-----------------+------------+-------------------+--------------
  1 | First milestone |          0 |                 1 |           50
(1 row)

; Run the UPDATE for the first time
giteatest=> UPDATE milestone SET num_issues=(SELECT count(*) FROM issue WHERE milestone_id=1),
    completeness = 100*num_closed_issues/(num_issues+1-sign(num_issues)) WHERE id=1;
UPDATE 1

; The UPDATE corrected the value for num_issues, but completeness was calculated using
; the previous value in num_issues
giteatest=> select id, name, num_issues, num_closed_issues, completeness
    from milestone where id = 1;
 id |      name       | num_issues | num_closed_issues | completeness
----+-----------------+------------+-------------------+--------------
  1 | First milestone |          2 |                 1 |          100
(1 row)

; Run the same UPDATE a second time.
giteatest=> UPDATE milestone SET num_issues=(SELECT count(*) FROM issue WHERE milestone_id=1),
    completeness = 100*num_closed_issues/(num_issues+1-sign(num_issues)) WHERE id=1;
UPDATE 1

; Now completeness is calculated correctly
; because num_issues already had the right value.
giteatest=> select id, name, num_issues, num_closed_issues, completeness
    from milestone where id = 1;
 id |      name       | num_issues | num_closed_issues | completeness
----+-----------------+------------+-------------------+--------------
  1 | First milestone |          2 |                 1 |           50
(1 row)
```

Reply:
I think your best bet is to use an UPDATE JOIN. The following works correctly in every case:

```
UPDATE milestone
SET num_issues=c.issues, num_closed_issues=c.closed_issues,
    completeness = 100*c.closed_issues/(c.issues+1-sign(c.issues))
FROM (SELECT count(*) as issues,
      SUM(CASE is_closed WHEN true THEN 1 ELSE 0 END) as closed_issues
      FROM issue WHERE milestone_id = ?) as c
WHERE id = ?;
```

Reply:
@guillep2k But it seems not all databases support the SQL well.

Reply:
Yeah, I see that sqlite3 can't handle that. 😞